### PR TITLE
[sparsity] Sparsifier class

### PR DESCRIPTION
**Pull request by z-a-f**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #60728 [quant][sparsity] Generic convert function
* #59771 [sparsity] Lambda Scheduler
* #59770 [sparsity] Base sparsity level scheduler class
* #58955 [sparsity] WeightNormSparsifier
* #58707 [sparsity][refactor] Import factoring out
* **#58704 [sparsity] Sparsifier class**
* #58705 [sparsity] Sparsity parametrization
* #60850 [sparsity][refactor] Changing linear row/col control
* #60887 [sparsity] Add sparsity tests to run_test.py

Implements the base sparsifier class based on the #59835 RFC documents.

This PR implements the base class for the sparsification. Specifically, the prepare method is implemented.

Test Plan:

```
python test/test_ao_sparsity.py
```

Differential Revision: [D28970958](https://our.internmc.facebook.com/intern/diff/D28970958)

### CODE REVIEW

**Comment by z-a-f** [comment]
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28970958).

**Comment by z-a-f** [comment]
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28970958).

File: test/ao/test_sparse_kernels.py [hunk 1]
```diff
@@ -156,6 +156,7 @@ def test_sparse_qlinear(self):
                 model.qconfig = tq.get_default_qconfig('fbgemm')
                 qmodel = copy.deepcopy(model)
                 sqmodel = copy.deepcopy(model)
+                sqmodel.linear.mask = mask
```
Comment (raghuramank100) [inline test/ao/test_sparse_kernels.py:159]: NIT: mask would be an attribute of the fake-sparse module

**Comment by z-a-f** [comment]
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28970958).

File: test/ao/sparsity/test_kernels.py [hunk 2]
```diff
@@ -156,6 +156,7 @@ def test_sparse_qlinear(self):
                 model.qconfig = tq.get_default_qconfig('fbgemm')
                 qmodel = copy.deepcopy(model)
                 sqmodel = copy.deepcopy(model)
+                sqmodel.linear.mask = mask
```
Comment (raghuramank100) [inline test/ao/sparsity/test_kernels.py:159]: Not sure what happened to previous comments: This should be sqmodel.linear.fake_sparse.mask

File: test/run_test.py [hunk 3]
```diff
@@ -33,6 +33,7 @@
     'test_import_time',
     'test_public_bindings',
     'test_type_hints',
+    'test_ao',
```
Comment (raghuramank100) [inline test/run_test.py:36]: Should this be test_ao_sparsity instead?

File: torch/ao/sparsity/_variables.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
+import torch
```
Comment (raghuramank100) [inline torch/ao/sparsity/_variables.py:1]: Suggest calling this sparsity_mappings.py, similar to: 
https://github.com/pytorch/pytorch/blob/master/torch/quantization/quantization_mappings.py

File: torch/ao/sparsity/_variables.py [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+def get_sparse_mapping():
```
Comment (raghuramank100) [inline torch/ao/sparsity/_variables.py:3]: Can these be plain dicts instead of functions, to align with quantization (see https://github.com/pytorch/pytorch/blob/master/torch/quantization/quantization_mappings.py#L72)

File: torch/ao/sparsity/sparsifier.py [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+from torch import nn
+from torch.nn.utils import parametrize
+
+from .parametrization import MulBy
+
+SUPPORTED_MODULES = {
+    nn.Linear
+}
+
+def _module_to_path(model, layer, prefix=''):
```
Comment (raghuramank100) [inline torch/ao/sparsity/sparsifier.py:15]: NIT: Suggest calling this _module_to_fqn instead.

File: torch/ao/sparsity/sparsifier.py [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+    for name, child in model.named_children():
+        new_name = prefix + '.' + name
+        if child is layer:
+            return new_name
+        child_path = _module_to_path(child, layer, prefix=new_name)
+        if child_path is not None:
+            return child_path
+    return None
+
+def _path_to_module(model, path):
```
Comment (raghuramank100) [inline torch/ao/sparsity/sparsifier.py:25]: Do both these functions work for sequentials? Also, for consistency suggest calling this _fqn_to_module

File: torch/ao/sparsity/sparsifier.py [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+        self.state = {}
+        self.module_groups = []
+        self.enable_mask_update = False
+
+        self.model = model
+        # If no config -- try getting all the supported layers
+        if self.config is None:
+            # Add all models to the config
+            self.config = []
+            stack = [model]
```
Comment (raghuramank100) [inline torch/ao/sparsity/sparsifier.py:74]: Why do we need a while loop here? Cant we just check for this with a for loop followed by an if?.

Also, in this case, we should have the user specify the list of modules, instead of passing the entire model

File: torch/ao/sparsity/sparsifier.py [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+            local_args = copy.deepcopy(self.defaults)
+            local_args.update(module_config)
+            module = local_args['module']
+            module_path = _module_to_path(self.model, module)
+            if module_path and module_path[0] == '.':
+                module_path = module_path[1:]
+            local_args['path'] = module_path
+            self.module_groups.append(local_args)
+
+    def __getstate__(self):
```
Comment (raghuramank100) [inline torch/ao/sparsity/sparsifier.py:95]: Do we intend for sparsifier objects to be pickleable? module groups contain references to modules, so they shouldnt be pickleable

File: torch/ao/sparsity/sparsifier.py [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+                raise RuntimeError(f'Error loading group["path"] into the model')
+            group['module'] = layer
+        self.__setstate__({'module_groups': module_groups})
+
+    def prepare(self, *args, **kwargs):
+        r"""Adds mask parametrization to the layer weight
+        """
+        for config in self.module_groups:
+            module = config['module']
+            if not getattr(module, 'mask', None):
```
Comment (raghuramank100) [inline torch/ao/sparsity/sparsifier.py:148]: Lets replace this with fakesparse module insertions, which should contain the mask

File: torch/ao/sparsity/sparsifier.py [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+                        self.config.append(child)
+                    else:
+                        stack.append(child)
+
+        for module_config in self.config:
+            if isinstance(module_config, nn.Module):
+                module_config = {'module': module_config}
+            local_args = copy.deepcopy(self.defaults)
+            local_args.update(module_config)
+            module = local_args['module']
```
Comment (raghuramank100) [inline torch/ao/sparsity/sparsifier.py:88]: module also needs to a reference to the weights so that step can use it.

File: torch/ao/sparsity/sparsifier.py [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+        self.config = config
+        self.defaults = defaults
+        if self.defaults is None:
+            self.defaults = dict()
+
+        self.state = {}
+        self.module_groups = []
+        self.enable_mask_update = False
+
+        self.model = model
```
Comment (raghuramank100) [inline torch/ao/sparsity/sparsifier.py:69]: Can we eliminate the need to have a model as an argument and instead do the following:

``` python
config = [ {'module': model.linear2, 'sparsity_level': 0.5}]
sparsifier = MySparsifier(config)
```

**Comment by z-a-f** [comment]
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28970958).

File: test/run_test.py (hunk 3 above)
Comment (z-a-f) [reply test/run_test.py]: This is weird -- we changed it with @kazhou, weirdly it is still here.

File: test/ao/sparsity/test_kernels.py (hunk 2 above)
Comment (z-a-f) [reply test/ao/sparsity/test_kernels.py]: That's not e en supposed to be tested here -- this bled through from some other PR

File: torch/ao/sparsity/_variables.py (hunk 5 above)
Comment (z-a-f) [reply torch/ao/sparsity/_variables.py:3]: The quantization_mappings also use functions. In fact, using regular dicts is a bad idea, as they are mutable

File: torch/ao/sparsity/sparsifier.py (hunk 9 above)
Comment (z-a-f) [reply torch/ao/sparsity/sparsifier.py:68]: At this stage the module_groups have only the FQNs, which we want to pickle, right?

File: torch/ao/sparsity/sparsifier.py [hunk 13]
```diff
@@ -0,0 +1,182 @@
...
+                        stack.append(child)
+
+        for module_config in self.config:
+            if isinstance(module_config, nn.Module):
+                module_config = {'module': module_config}
+            local_args = copy.deepcopy(self.defaults)
+            local_args.update(module_config)
+            module = local_args['module']
+            module_path = _module_to_fqn(self.model, module)
+            if module_path and module_path[0] == '.':
```
Comment (raghuramank100) [inline torch/ao/sparsity/sparsifier.py:146]: Can we handle this logic in _module_to_fqn?

File: torch/ao/sparsity/sparsifier.py [hunk 14]
```diff
@@ -0,0 +1,182 @@
...
+            self.module_groups.append(local_args)
+
+        self._prepare()
+
+    def _prepare(self, *args, **kwargs):
+        r"""Adds mask parametrization to the layer weight
+        """
+        for config in self.module_groups:
+            module = config['module']
+            if getattr(module, 'mask', None) is None:
```
Comment (raghuramank100) [inline torch/ao/sparsity/sparsifier.py:158]: Can we have masks be only attr of FakeSparsity? If we change FakeSparsity's init to only take the weight shape, this is simpler. Currently there are two masks: one as an attr of the module and one as part of the FakeSparsityModule.
Suggest instead doing:
``` python
for config in self.module_groups:
            module = config['module']
            param = config.get('parametrization', FakeSparsity)
            parametrize.register_parametrization(module, 'weight',
                                                 param(getattr(module,'weight').shape))
```

Comment (z-a-f) [reply torch/ao/sparsity/sparsifier.py:158]: Oh the intention was that "if there is a mask in the module, it would use that, otherwise it would create one

File: torch/ao/sparsity/sparsifier.py [hunk 15]
```diff
@@ -0,0 +1,182 @@
...
+            param = config.get('parametrization', FakeSparsity)
+            parametrize.register_parametrization(module, 'weight',
+                                                 param(module.mask))
+
+    def squash_mask(self, *args, **kwargs):
+        for config in self.module_groups:
+            module = config['module']
+            parametrize.remove_parametrizations(module, 'weight',
+                                                leave_parametrized=True)
+            if getattr(module._parameters, 'mask', None):
```
[the remark this answers is not in the document]
Comment (z-a-f) [reply torch/ao/sparsity/sparsifier.py:169]: Because this is a generalized approach, we must consider a case where the mask is part of the module. I will add a check if there is an attribute in the module

File: torch/ao/sparsity/sparsifier.py (hunk 9 above)
Comment (raghuramank100) [reply torch/ao/sparsity/sparsifier.py:68]: Yes, that is correct.

File: torch/ao/sparsity/sparsifier.py [hunk 16]
```diff
@@ -0,0 +1,182 @@
...
+            format_string += f'\tModule Group {i}\n'
+            format_string += f'\t    module: {module}\n'
+            for key in sorted(sparse_args.keys()):
+                if key == 'module':
+                    continue
+                format_string += f'\t    {key}: {sparse_args[key]}\n'
+        format_string += ')'
+        return format_string
+
+    def state_dict(self):
```
Comment (raghuramank100) [inline torch/ao/sparsity/sparsifier.py:92]: This is how I think of the state:
```python
# state returned by getstate
 __getstate__(self):
  return {
            'defaults': self.defaults,
            'state': self.state,
            'module_groups': self.module_groups,
        }
# state_dict
 # {'fqn0': {'name1': Tensor1, 'name2': Tensor2, ....},
     'fqn1': {'name1':Tensor3, 'name2': Tensor4,...}
   }
 # This gets assigned to self.state

File: torch/ao/sparsity/sparsifier.py (hunk 14 above)
Comment (raghuramank100) [reply torch/ao/sparsity/sparsifier.py:158]: Lets make it explicit, the module should not have a mask, this way mask is always part of fakesparse module.

**Comment by z-a-f** [comment]
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28970958).

**Comment by z-a-f** [comment]
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28970958).

File: torch/ao/sparsity/sparsifier.py (hunk 13 above)
Comment (z-a-f) [reply torch/ao/sparsity/sparsifier.py:146]: I don't think with the current implementation it is that simple -- the name having an empty root name is only resolved when the _module_to_fqn exits all the recursion calls. To make it part of it, I would need to create additional function(wrapper) around the _module_to_fqn.

**Comment by z-a-f** [comment]
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28970958).

File: test/ao/sparsity/test_sparsifier.py [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+        assert sparsifier.module_groups[0]['fqn'] == 'linear'
+        assert 'test' in sparsifier.module_groups[0]
+        assert sparsifier.module_groups[0]['test'] == 3
+
+    def test_state_dict(self):
+        model = Model()
+        sparsifier0 = ImplementedSparsifier(test=3)
+        sparsifier0.prepare(model, [model.linear])
+        state_dict = sparsifier0.state_dict()
+        sparsifier1 = ImplementedSparsifier()
```
Comment (raghuramank100) [inline test/ao/sparsity/test_sparsifier.py:53]: Can we add an explicit check for what the keys and values of the state_dict should look like here?

File: torch/ao/sparsity/_mappings.py [hunk 18]
```diff
@@ -0,0 +1,25 @@
...
+
+def get_sparse_mapping():
+    _sparse_mapping = dict({
+        torch.nn.Linear: torch.ao.nn.sparse.Linear,
+    })
+    return _sparse_mapping
+
+def get_static_sparse_quantized_mapping():
+    _static_sparse_quantized_mapping = dict({
+        # Dense
```
Comment (raghuramank100) [inline torch/ao/sparsity/_mappings.py:11]: Why are the dense mappings present here? Lets make this simpler: We only map from sparse.Linear to the quantized Linear modules. If someone wants to override this, they can.

File: torch/ao/sparsity/sparsifier.py [hunk 19]
```diff
@@ -0,0 +1,213 @@
...
+                if isinstance(p, FakeSparsity):
+                    parametrization = p
+                    found = True
+                    break
+            if found:
+                mask = parametrization.mask
+            state[original_weight]['mask'] = mask
+            state[original_weight]['fqn'] = g['fqn']
+            # Get all the tensors inside the module_group
+            state[original_weight].update(
```
Comment (raghuramank100) [inline torch/ao/sparsity/sparsifier.py:111]: This doesnt look right. We cannot serialize the original weight as a key to the state.

File: torch/ao/sparsity/sparsifier.py [hunk 20]
```diff
@@ -0,0 +1,213 @@
...
+                for p in layer.parametrizations['weight']:
+                    if isinstance(p, FakeSparsity):
+                        found = True
+                        break
+                if not found:
+                    p = FakeSparsity(torch.ones(group['module'].weight.shape))
+                    parametrize.register_parametrization(layer, 'weight', FakeSparsity(p))
+
+            group['module'] = layer
+        self.__setstate__({'module_groups': module_groups})
```
Comment (raghuramank100) [inline torch/ao/sparsity/sparsifier.py:152]: What about restoring the 'state' ?

File: torch/ao/sparsity/sparsifier.py (hunk 19 above)
Comment (z-a-f) [reply torch/ao/sparsity/sparsifier.py:111]: That what the optimizer does though -- I am mimicking what they do

**Comment by z-a-f** [comment]
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28970958).

File: torch/ao/sparsity/sparsifier.py (hunk 19 above)
Comment (z-a-f) [reply torch/ao/sparsity/sparsifier.py:111]: But I think using fqns as keys should be fine.

File: torch/ao/sparsity/sparsifier.py (hunk 20 above)
Comment (z-a-f) [reply torch/ao/sparsity/sparsifier.py:152]: Forgot about it :/

**Comment by z-a-f** [comment]
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28970958).

**Review by raghuramank100 — APPROVED** [review]
Please verify tests with latest changes prior to landing

**Comment by z-a-f** [comment]
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28970958).

**Comment by z-a-f** [comment]
@zafartahirov has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28970958).
